feat: Make pip installable package

matthewfeickert opened the pull request:
This PR turns the repo into a `pip` installable Python 3 package called `ads2inspire` that can be installed with all its dependencies directly from the repo (in editable mode `-e`) with

```
python -m pip install -e .
```

This is done by using both a modern `src` directory layout with a corresponding `setup.py`, `setup.cfg`, and `pyproject.toml`. The `setup.py` contains all the logic for installation, while all the metadata and options that do not require logic are in `setup.cfg`. `pyproject.toml` provides the ability to do modern PEP 517/518 style builds of the src and wheel distributions. 

The dependencies that would be needed for testing (upcoming PR) and development have been added in an additional `test` and `develop` "extra" that can be all installed through

```
python -m pip install -e .[test] # here demoing for test
```

or to get all additional extra dependencies

```
python -m pip install -e .[complete]
```

Almost no code is really changed in this PR &mdash; just moved around and refactored lightly. The only major change of code is the use of [`Click`](https://click.palletsprojects.com/en/7.x/) over `argparse` to much more easily create CLI APIs.

Also resolves #1

matthewfeickert committed:
Use a src directory structure for library

Split the tooling across the init file and an CLI file

matthewfeickert committed:
Add files for installation and dist builds

setup.py and setup.cfg give all information for installation of the
package. pyproject.toml allows for modern builds of the distributions.

matthewfeickert committed:
Add MANIFEST for use in packaging

matthewfeickert committed:
Add installation instructions to README

matthewfeickert commented:
@duetosymmetry This is ready for review. I can imagine that as this is a lot of restructure that you'll have questions, and I'm more then happy to try to answer any of them. :)

I've run some basic testing of my own, and can add CI tests in another PR, but if you have specific test cases you want run let me know.

duetosymmetry commented on src/ads2inspire/__init__.py:
```diff
@@ -0,0 +1,231 @@
+from pathlib import Path
+import argparse
```
argparse is no longer needed

matthewfeickert replied:
Good catch!

duetosymmetry commented on setup.cfg:
```diff
@@ -0,0 +1,50 @@
+[metadata]
+name = ads2inspire
+version = 0.1.0
```
Should I be using semantic or calendar versioning?
Should the source have access to the version string?

matthewfeickert replied:
> Should I be using semantic or calendar versioning?

This is a debate that rages. I use semantic versioning for all my projects, but I have my own biases. I can send you videos that argue for calendar versioning if you want to hear some other agruments.

> Should the source have access to the version string?

We do this for `pyhf`, but this can be added in another PR. Though if you'd prefer to have that in now that's not a problem.

duetosymmetry commented on setup.cfg:
```diff
@@ -0,0 +1,50 @@
+[options.packages.find]
+where = src
+
+[options.entry_points]
+console_scripts =
+    ads2inspire = ads2inspire.cli:ads2inspire
```
This and the `click` usage is so neat!

duetosymmetry reviewed (changes requested):
This is so neat, thanks for your help and for educating me at the same time! Only minimal changes requested. I'm adding better docstrings right now, will rebase onto your commits after merge.

matthewfeickert committed:
Remove argparse to pass pyflakes

To get pyflakes to be happy, remove the use of argparse and then use
f-strings for best practices.

matthewfeickert commented:
> I'm adding better docstrings right now, will rebase onto your commits after merge.

Yeah! Good docs FTW! :rocket: I'm sorry though, as given the changes it will probably require you to manually copy and paste the changes over. :/

duetosymmetry reviewed (commented):
LGTM
